### PR TITLE
fix($location): avoid unnecessary `$locationChange*` events due to empty hash

### PR DESCRIPTION
**Pull request by gkalpak**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Fix (among others).

**What is the current behavior? (You can also link to an open issue here)**
#16632

**What is the new behavior (if this is a feature change)?**
No unnecessary `$locationChange*` events due to empty hash

**Does this PR introduce a breaking change?**
Hopefully not :grin:

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] ~~Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated~~
- [x] Fix/Feature: Tests have been added; existing tests pass

### CODE REVIEW

File: src/ng/browser.js [hunk 1]
```diff
@@ -143,7 +143,7 @@ function Browser(window, document, $log, $sniffer, $$taskTrackerFactory) {
       // - pendingLocation is needed as browsers don't allow to read out
       //   the new location.href if a reload happened or if there is a bug like in iOS 9 (see
       //   https://openradar.appspot.com/22186109).
-      return pendingLocation || location.href;
+      return pendingLocation || trimEmptyHash(location.href);
```
Comment (jbedard) [inline src/ng/browser.js]: What about `pendingLocation`?

**Comment by jbedard** [comment]
Where is `$browser.url()` being used to cause this bug? I question if we should fix it there instead similar to aee7d53a6b5d3d7bc0a1124fd3df9b263777e72e

File: src/ng/browser.js (hunk 1 above)
Comment (gkalpak) [reply src/ng/browser.js]: `pendingLocation` always has the empty hash fragment stripped. In a nutshell, `pendingLocation` is always the value `$browser.url(url)` is called with, which is always `$location.absUrl()` (directly or indirectly), which always has the empty hash stripped off.

This potential difference between `pendingLocation` and `location.href` is what causes the issue. Because while `pendingLocation` is defined, `$browser.url()` return the url without the empty hash, but once `pendingLocation` gets undefined (i.e. on `hashchange`, `popstate`, which may be triggered asynchronously), then `$browser.url()` starts returning the url with the empty fragment.

So, the returned url changes, even if there is no actual change in the url.

**Comment by gkalpak** [comment]
> Where is `$browser.url()` being used to cause this bug? I question if we should fix it there instead similar to aee7d53

[This is one place][1]. There may be others.
Note that is many places we trim the empty hash before using the returned url. (If we decide this fix is desired, I will remove those unnecessary calls.)

Also, this looks a little problematic:

https://github.com/angular/angular.js/blob/6706353a71e3c11c56c0b19be03600dac57aafe1/src/ng/location.js#L970-L979 

Although we don't trim the empty hash for determining there's been a url change, we trim it before broadcasting the event. So, you will get a `$locationChangeStart` event from `/foo/` to `/foo/` 😱

##
In general, I prefer to address the root of the problem. Patching stuff "just-in-time" to avoid a specific bug makes the code too complicated and hard to follow.

(But I understand `$location`/`$browser` might a special case.)

[1]: https://github.com/angular/angular.js/blob/6706353a71e3c11c56c0b19be03600dac57aafe1/src/ng/browser.js#L192

**Comment by jbedard** [comment]
While patching stuff "just-in-time" can be iffy, modifying something generic like `$browser.url()` for a specific case that might not apply everywhere is also iffy. I think that was one reason we chose not to modify `$browser.url()` in aee7d53a6b5d3d7bc0a1124fd3df9b263777e72e?

For the `$locationChangeStart` `newUrl` thing... can we just drop the `newUrl = trimEmptyHash(newUrl);` line?

**Comment by gkalpak** [comment]
> modifying something generic like $browser.url() for a specific case that might not apply everywhere is also iffy.

My point is that the proposed change _does_ apply everywhere and does not change the behavior of `$browser` other than not incorrectly reporting URL changes. 

But I will understand if you feel uncomfortable with changing `.url()` :smiley:

**Comment by jbedard** [comment]
But is adding/removing `#` a URL change? The browser thinks it is... 🤔

File: src/ng/browser.js (hunk 1 above)
Comment (jbedard) [reply src/ng/browser.js]: > This potential difference between pendingLocation and location.href is what causes the issue

Aha, that helps me understand this a bit more. Why does `location.href` have the hash in the first place?

Comment (gkalpak) [reply src/ng/browser.js]: That's what browsers do (sometimes :grin:). I found that this can happen when clicking a link with a trailing # (but not when directly setting `location.hash`).

**Comment by gkalpak** [comment]
I think you already found https://github.com/angular/angular.js/pull/16636#discussion_r203808513, which (kind of) explains that it's not the browser that thinks there's a change, it's `$browser.url()` that temporarily thinks so (while `pendingLocation` is set).

**Comment by jbedard** [comment]
I mean if you append `#` to the end of a URL and press enter, the browser will push a new entry into the history. So normally browsers do consider `/` and `/#` to be different URLs.

I guess I just want to acknowledge that the browser does consider them different, it's only us/`$location` that treats them the same. IDK if that actually means anything though >_>

I'm curious what other code could be simplified with the current fix... and how the `$locationChangeStart` event can be fixed. Want to go ahead with this as-is for now? I'm still tempted to try the other method and compare though...

File: src/ng/browser.js (hunk 1 above)
Comment (jbedard) [reply src/ng/browser.js]: I think I'd vote to still do `trimEmptyHash(pendingLocation || location.href)` just for consistency...

File: src/ng/location.js [hunk 2]
```diff
@@ -93,7 +94,7 @@ function stripHash(url) {
 }
 
 function trimEmptyHash(url) {
```
Comment (jbedard) [inline src/ng/location.js]: Can we remove some calls to this within this file now that `$browser` does it?

Comment (gkalpak) [reply src/ng/location.js]: [Absolutely!][1]

> Note that is many places we trim the empty hash before using the returned url. (If we decide this fix is desired, I will remove those unnecessary calls.)

[1]: https://github.com/angular/angular.js/pull/16636#issuecomment-406356498

File: src/ng/browser.js (hunk 1 above)
Comment (gkalpak) [reply src/ng/browser.js]: Done.

File: src/ng/browser.js [hunk 3]
```diff
@@ -1,5 +1,14 @@
 'use strict';
-/* global stripHash: true */
+/* global getHash: true, stripHash: false, trimEmptyHash: true */
```
Comment (petebacondarwin) [inline src/ng/browser.js]: where is `getHash` used globally?

**Review by petebacondarwin — APPROVED** [review]
Couple of minor comments.

File: src/ng/browser.js (hunk 3 above)
Comment (gkalpak) [reply src/ng/browser.js]: It is "used" in tests, in a mock window. The feature that relies on it is not actually used atm, so there were no breakages.

See https://github.com/angular/angular.js/pull/16636/commits/34fe24f76e4b09c1a6574518eb6402a7a06c03ef (yes, there is a typo in that commit message :grin:).

File: src/ng/location.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
 'use strict';
+/* global stripHash: true */
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ng/location.js]: `stripHash` is used more in `$location` than `$browser` and other similar helpers (e.g. `stripBaseUrl`, `stripFile`) are also in `$location`. I only moved `trimEmptyHash`, because it is only used in `$browser` now.

Even if we had imports/exports (which we don't because everything lives inside the great AngularJS IIFE :grin:), I don't see two way importing going on. `$location` would export `stripHash` and `$browser` would import `stripHash` and export `getHash` and `trimEmptyHash`.

(We definitely have enough URL-specific helpers that we could move them into a separate file and "import" from there :grin:)

File: test/ng/locationSpec.js [hunk 5]
```diff
@@ -707,10 +707,11 @@ describe('$location', function() {
 
     it('should not replace browser url if only the empty hash fragment is cleared', function() {
       initService({html5Mode: true, supportHistory: true});
-      mockUpBrowser({initialUrl:'http://new.com/#', baseHref: '/'});
-      inject(function($browser, $location) {
-        expect($browser.url()).toBe('http://new.com/#');
+      mockUpBrowser({initialUrl: 'http://new.com/#', baseHref: '/'});
+      inject(function($browser, $location, $window) {
+        expect($browser.url()).toBe('http://new.com/');
```
Comment (Narretz) [inline test/ng/locationSpec.js]: The (non-public) docs for browser.url() say "Without any argument, this method just returns current value of location.href." Isn't this now wrong as the test shows a difference between url() and location.href?

Comment (gkalpak) [reply test/ng/locationSpec.js]: Good point.

File: src/ng/location.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbedard) [reply src/ng/location.js]: I think it's a little weird having `$browser` depend on `$location`. Or was that always the case?

**Review by jbedard — APPROVED** [review]
